consensus/state.go

```diff
@@ -1339,55 +1339,59 @@ func (cs *State) defaultDoPrevote(height int64, round int32) {
-	if !isAppValid {
-		logger.Error("prevote step: state machine rejected a proposed block; this should not happen:"+
-			"the proposer may be misbehaving; prevoting nil", "err", err)
-		cs.signAddVote(cmtproto.PrevoteType, nil, types.PartSetHeader{})
-		return
-	}
-
 	/*
 		22: upon <PROPOSAL, h_p, round_p, v, −1> from proposer(h_p, round_p) while step_p = propose do
 		23: if valid(v) && (lockedRound_p = −1 || lockedValue_p = v) then
 		24:   broadcast <PREVOTE, h_p, round_p, id(v)>
 		25: else
 		26:   broadcast <PREVOTE, h_p, round_p, nil>
 
-		Here, cs.Proposal.POLRound corresponds to the -1 in the above algorithm rule.
+		Here, cs.Proposal.POLRound corresponds to the -1 in the rule of the pseude-code (line 22).
 		This means that the proposer is producing a new proposal that has not previously
 		seen a 2/3 majority by the network.
 
-		If we have already locked on a different value that is different from the proposed value,
-		we prevote nil since we are locked on a different value. Otherwise, if we're not locked on a block
-		or the proposal matches our locked block, we prevote the proposal.
+		If the application deems the proposal as valid AND we're not locked on a
+		block OR the proposal matches our locked block (line 23), we prevote the
+		proposal (line 24). Note that when the proposal matches our locked
+		block, we do not need to query the application to validate the proposal,
+		because we know it was validated by at least by one correct node the
+		moment we locked it.
+
+		Otherwise, we have already locked on a value that is different from the
+		proposed value, so we prevote nil (line 26), because we are locked on a
+		different value.
 	*/
 	if cs.Proposal.POLRound == -1 {
-		if cs.LockedRound == -1 {
-			logger.Debug("prevote step: ProposalBlock is valid and there is no locked block; prevoting the proposal")
+		if cs.ProposalBlock.HashesTo(cs.LockedBlock.Hash()) {
+			logger.Debug("prevote step: ProposalBlock is valid (POLRound is -1) and matches our locked block; prevoting the proposal")
 			cs.signAddVote(cmtproto.PrevoteType, cs.ProposalBlock.Hash(), cs.ProposalBlockParts.Header())
 			return
 		}
-		if cs.ProposalBlock.HashesTo(cs.LockedBlock.Hash()) {
-			logger.Debug("prevote step: ProposalBlock is valid (POLRound is -1) and matches our locked block; prevoting the proposal")
+
+		// Before prevoting on the block received from the proposer for the current round and height,
+		// we request the Application, via `ProcessProposal` ABCI call, to confirm that the block is
+		// valid. If the Application does not accept the block, consensus prevotes nil.
+		//
+		// WARNING: misuse of block rejection by the Application can seriously compromise
+		// the liveness properties of consensus.
+		// Please see `PrepareProosal`-`ProcessProposal` coherence and determinism properties
+		// in the ABCI++ specification.
+		isAppValid, err := cs.blockExec.ProcessProposal(cs.ProposalBlock, cs.state)
```

Comment:
Here we are calling `ProcessProposal`, but if `cs.LockedRound == -1` is false, we will be prevoting `nil`

Reply:
That's right. It's fixed with the above suggestion.